languagetool-core/src/main/java/org/languagetool/RuleEntityResolver.java

```diff
@@ -26,22 +23,19 @@ public class RuleEntityResolver implements EntityResolver {
    */
   @Override
   public InputSource resolveEntity(String publicId, String systemId) throws SAXException, IOException {
-    if ((publicId != null && publicId.endsWith(".ent")) || systemId.endsWith(".ent")) {
-      URL entitiesUrl = new URL(xmlUrl, new URL(systemId).getPath());
-      if (Objects.equals(entitiesUrl.getProtocol(), "jar")) {
-        InputStream entitiesStream = JLanguageTool.getDataBroker().getAsStream(entitiesUrl.toString().split("!")[1]);
-        return new InputSource(entitiesStream);
-      } else {
-        return new InputSource(entitiesUrl.getPath());
-      }
+    if (systemId != null && systemId.endsWith(".ent")) {
+      return new InputSource(getInputStreamLTEntities(systemId));
     }
     return null;
   }
 
-  /**
-   * @param xmlUrl path to the source XML where the external entity relative path is found.
-   */
-  public RuleEntityResolver(URL xmlUrl) {
-    this.xmlUrl = xmlUrl;
+  public String getPathFromLTResourceFolder(String input) {
+    // we assume that the entities file is in the resource folder of each language
+    return input.replaceAll(".*/resource/", "").replaceAll("\\.\\./", "");
```

Comment:
What if a grammar file is inside a further directory (e.g. `pt/pt-BR`). There, the relative path will be written `../../../`. Will the `.replaceAll()`s here be enough?

Reply:
It is enough. In most cases, `.replaceAll(".*/resource/", "")` will be enough. In fact, it is a requirement to have this string in the relative path.